sys/universal_address/universal_address.c

```diff
@@ -196,41 +196,47 @@ int universal_address_compare(universal_address_container_t *entry,
         return ret;
     }
 
-    /* Get the index of the first trailing `0` (indicates a prefix) */
-    int i = 0;
-    for( i = entry->address_size-1; i > 0; --i) {
-        if( entry->address[i] != 0 ) {
+    /* compare up to fist distinct byte, pretty clumsy method for now */
+    int idx = -1;
+    bool test_all_zeros = true;
+    for( size_t i = 0; i < entry->address_size; i++) {
+        if ( (idx == -1) && (entry->address[i] != addr[i]) ) {
+            idx = i;
+        }
+        if ( test_all_zeros ) {
+            test_all_zeros = (entry->address[i] == 0);
+        }
+        if ( (idx != -1) && !test_all_zeros ) {
             break;
         }
     }
 
-    if( memcmp(entry->address, addr, i) == 0 ) {
-        /* if the bytes-1 equals we check the bits of the lowest byte */
-        uint8_t bitmask = 0x00;
-        uint8_t j = 0;
-        /* get a bitmask for the trailing 0b */
-        for( ; j < 8; ++j ) {
-            if ( (entry->address[i] >> j) & 0x01 ) {
-                bitmask = 0xff << j;
-                break;
-            }
-        }
-        if( (entry->address[i] & bitmask) == (addr[i] & bitmask) ) {
-            ret = entry->address[i] != addr[i];
-            *addr_size_in_bits = (i<<3) + (8-j);
-            if( ret == 0 ) {
-                /* check if the remaining bits from addr are significant */
-                i++;
-                for(; i < entry->address_size; ++i) {
-                    if( addr[i] != 0 ) {
-                        ret = 1;
-                        break;
-                    }
-                }
-            }
+    /* if the address is all 0 its a default route address */
+    if ( test_all_zeros ) {
+        *addr_size_in_bits = 0;
+        mutex_unlock(&mtx_access);
+        return 2;
```

Comment:
I would prefer constant instead of a literal here (and the doxygen needs updating)